macps.1
.\" SCCSid = "@(#)macps.1	2.2 10/24/89"
.TH MACPS 1 "24 Oct 1989"
.UC 4
.SH NAME
macps \- print Macintosh-created PostScript file on Unix
.SH SYNOPSIS
.B macps
[
\-c copies
]
[
\-d directory
]
[
\-r
]
[
file
]
.SH DESCRIPTION
.I Macps
takes the command-line file (or the standard input if no filename is given),
and prepends the appropriate Macintosh LaserPrep file (e.g., those generated
by
.IR prepfix (1)).
The standard output can then be directed to a PostScript printer, via
.IR lpr (1),
for example.
.PP
The input PostScript file is generated on a Macintosh by typing (and holding)
Command-F immediately after clicking the OK button of the LaserWriter printer
dialog box.
Another dialog will appear, confirming that a file named ``PostScript'' is
being created (the Command-F keys can be released now).
.PP
Depending on the application, the created PostScript file can be found in the
System Folder, the application folder or the top level of the disk.
This file can then be uploaded via some file transfer program, such as
MacTerminal/VersaTerm and
.IR macget (1)
or MacKermit/Red Ryder and
.IR kermit (1).
.PP
Normally, you would specify the number of copies in the LaserWriter
print dialog box on the Macintosh.
However, you can override that after uploading the PostScript file by
specifying the
.B \-c
option, followed by the number of copies.
.PP
The file
.B macps.config
specifies the mapping between the internal LaserPrep name and the actual
file it resides in.
This file is normally located in the lib subdirectory from which
.I macps
was called from.
The
.B \-d
option allow you to specify an alternate directory in which the
.B macps.config
file can be found.
.PP
Most Macintosh applications produce normal ASCII PostScript files.
However, some graphics programs that manipulate gray-scale images will
produce PostScript files that contain 8-bit binary data.
Since
.I macps
normally converts this binary data, these PostScript files will not work
properly.
The
.B \-r
(raw) option suppresses this binary conversion.
(Note: Depending on how the printer is physically connected, it may not be
able to handle 8-bit binary data properly, and results may be disappointing.)
.SH FILES
.TP "\w'lib/macps.config   'u"
lib/macps.config
maps ProcSet ids to LaserPrep filenames
.SH "SEE ALSO"
lpr(1), macget(1), kermit(1), prepfix(1)
.SH BUGS
.I Macps
only works with version 4.0 and up of the Macintosh LaserPrep files.
Because of the way bit smoothing is implimented by the LaserWriter driver,
some PostScript printers other than the Apple LaserWriters may not be able to
bit smooth.


macps.c
/*
 * Copyright (c) 1988, The Regents of the University of California.
 * Edward Moy, Workstation Software Support Group, Workstation Support Serices,
 * Information Systems and Technology.
 *
 * Permission is granted to any individual or institution to use, copy,
 * or redistribute this software so long as it is not sold for profit,
 * provided that this notice and the original copyright notices are
 * retained.  The University of California makes no representations about the
 * suitability of this software for any purpose.  It is provided "as is"
 * without express or implied warranty.
 */

#ifndef lint
static char *SCCSid = "@(#)macps.c	2.2 10/25/89";
#endif

#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/file.h>
#include <fcntl.h>
#include <time.h>
#include <unistd.h>
#include "str.h"
#include "ucbwhich.h"

#define	CONFIG		"macps.config"
#ifdef SYSV
#define	index		strchr
#define	rindex		strrchr
#endif

#ifdef SAVE
char *finale = "clear countdictstack 2 sub{end}repeat macps restore\n";
char intro[] = "\
%%! *** Created by macps: %s\
/macps save def\n\
";
#else
char intro[] = "\
%%! *** Created by macps: %s\
";
#endif
char *myname;
int ncopies = 0;
#ifdef CONFIGDIR
char ucblib[UCBMAXPATHLEN] = CONFIGDIR;
#else
int ucbalternate;
char ucbpath[UCBMAXPATHLEN];
char ucblib[UCBMAXPATHLEN];
#endif

void Usage();

int main(argc, argv)
int argc;
char **argv;
{
	register STR* str;
	register char *cp, *pp;
	register FILE *fp;
	register int i, fd;
	char line[BUFSIZ];
	char path[UCBMAXPATHLEN];
	long ltime;
	char *ctime();

#ifndef CONFIGDIR
	ucbwhich(*argv);
#endif
	if((myname = rindex(*argv, '/')))
		myname++;
	else
		myname = *argv;
	cp = NULL;
	for(argc--, argv++ ; argc > 0 && **argv == '-' ; argc--, argv++) {
		switch((*argv)[1]) {
		 case 'c':	/* multiple copies */
			if((*argv)[2])
				ncopies = atoi(&(*argv[2]));
			else {
				if(argc < 2)
					Usage();	/* never returns */
				argc--;
				ncopies = atoi(*++argv);
			}
			if(ncopies <= 0)
				Usage();	/* never returns */
			break;
		 case 'd':	/* alternate directory for config file */
			if((*argv)[2])
				cp = &(*argv[2]);
			else {
				if(argc < 2)
					Usage();	/* never returns */
				argc--;
				cp = *++argv;
			}
			strcpy(ucblib, cp);
			break;
		 case 'r':	/* raw mode */
			rawmode++;
			break;
		 default:
			Usage();	/* never returns */
		}
	}
	if(argc > 1)
		Usage();	/* never returns */
	if(argc == 1 && freopen(*argv, "r", stdin) == NULL) {
		fprintf(stderr, "%s: can't open %s\n", myname, *argv);
		exit(1);
	}
	str = STRalloc();
	if(!STRgets(str, stdin)) {
		fprintf(stderr, "%s: Null input\n", myname);
		exit(1);
	}
	strcat(ucblib, "/");
	strcpy(path, ucblib);
	strcat(path, CONFIG);
	if((fp = fopen(path, "r")) == NULL) {
		fprintf(stderr, "%s: Can't open %s\n", myname, path);
		exit(1);
	}
	time(&ltime);
	printf(intro, ctime(&ltime));
	do {
		if(ncopies != 0 && STRheadcompare(str, "userdict /#copies ")
		 == 0)
			continue;
		if(STRcompare(str, "%%EOF\n") == 0) {
#ifdef SAVE
			if(finale) {
				fputs(finale, stdout);
				finale = NULL;
			}
#endif
			STRputs(str, stdout);
			continue;
		}
		if(STRheadcompare(str, "%%IncludeProcSet:") == 0) {
			for(cp = (char *)&str->bufptr[17] ; ; cp++) {
				if(!*cp) {
					fprintf(stderr,
				 "%s: Syntax error on IncludeProcSet line\n",
					 myname);
					exit(1);
				}
				if(!isascii(*cp) || !isspace(*cp))
					break;
			}
			pp = (char *)str->curendptr;
			while(--pp >= cp) {
				if(!isascii(*pp) || !isspace(*pp))
					break;
				*pp = 0;
			}
			str->curendptr = (unsigned char *)(pp + 1);
			fseek(fp, 0L, 0);
			for( ; ; ) {
				if(!fgets(line, BUFSIZ, fp)) {
					fprintf(stderr,
					 "%s: Unknown IncludeProcSet %s\n",
					 myname, cp);
					exit(1);
				}
				if(*line == '#')
					continue;
				if((pp = index(line, '\n'))) {
					if(pp == line)
						continue;
					*pp = 0;
				}
				if(!(pp = index(line, '\t'))) {
					fprintf(stderr,
					 "%s: Syntax error in macps.config\n",
					 myname);
					exit(1);
				}
				*pp++ = 0;
				if(STRcompareptr(str, cp, line) == 0)
					break;
			}
			if(*pp == '/')
				strcpy(path, pp);
			else {
				strcpy(path, ucblib);
				strcat(path, pp);
			}
			fflush(stdout);
			if((fd = open(path, O_RDONLY, 0)) < 0) {
				fprintf(stderr, "%s: Can't open %s\n", myname,
				 path);
				exit(1);
			}
			while((i = read(fd, line, BUFSIZ)) > 0)
				write(1, line, i);
			close(fd);
			continue;
		}
		STRputs(str, stdout);
		if(ncopies > 1 && isascii(*str->bufptr) &&
		 isdigit(*str->bufptr)) {
			cp = (char *)str->bufptr;
			while(cp < (char *)str->curendptr && isascii(*cp)
			 && isdigit(*cp))
				cp++;
			if((char *)str->curendptr - cp == 4 &&
			 STRcompareptr(str, cp, " mf\n") == 0) {
				printf("userdict /#copies %d put\n", ncopies);
				ncopies = -1;
			}
		}
	} while(STRgets(str, stdin));
#ifdef SAVE
	if(finale)
		fputs(finale, stdout);
#endif
	exit(0);
}

void
Usage()
{
	fputs("Usage: macps [-c #] [-d directory] [-r] [file]\n", stderr);
	exit(1);
}
